common/scala/src/main/scala/whisk/core/entity/ActivationStore.scala

```diff
@@ -0,0 +1,126 @@
+
+import akka.actor.ActorSystem
+import akka.stream.ActorMaterializer
+import spray.json.JsObject
+import whisk.common.{Logging, TransactionId}
+import whisk.core.database.CacheChangeNotification
+import whisk.spi.Spi
+
+import scala.concurrent.Future
+
+trait ActivationStore {
+
+  /**
+   * Stores an activation.
+   *
+   * @param activation activation to store
+   * @param transid transaction ID for request
+   * @param notifier cache change notifier
+   * @return Future containing DocInfo related to stored activation
+   */
+  def store(activation: WhiskActivation)(implicit transid: TransactionId,
+                                         notifier: Option[CacheChangeNotification]): Future[DocInfo]
+
+  /**
+   * Retrieves an activation corresponding to the specified activation ID.
+   *
+   * @param activationId ID of activation to retrieve
+   * @param transid transaction ID for request
+   * @return Future containing the retrieved WhiskActivation
+   */
+  def get(activationId: ActivationId)(implicit transid: TransactionId): Future[WhiskActivation]
+
+  /**
+   * Deletes an activation corresponding to the provided activation ID.
+   *
+   * @param activationId ID of activation to delete
+   * @param transid transaction ID for the request
+   * @param notifier cache change notifier
+   * @return Future containing a Boolean value indication whether the activation was deleted
+   */
+  def delete(activationId: ActivationId)(implicit transid: TransactionId,
+                                         notifier: Option[CacheChangeNotification]): Future[Boolean]
+
+  /**
+   * Counts the number of activations in a namespace.
+   *
+   * @param namespace namespace to query
+   * @param name entity name to query
+   * @param skip number of activations to skip
+   * @param since timestamp to retrieve activations after
+   * @param upto timestamp to retrieve activations before
+   * @param transid transaction ID for request
+   * @return Future containing number of activations returned from query in JSON format
+   */
+  def countActivationsInNamespace(namespace: EntityPath,
+                                  name: Option[EntityPath] = None,
+                                  skip: Int,
+                                  since: Option[Instant] = None,
+                                  upto: Option[Instant] = None)(implicit transid: TransactionId): Future[JsObject]
+
+  /**
+   * Returns activations corresponding to provided entity name.
+   *
+   * @param namespace namespace to query
+   * @param name entity name to query
+   * @param skip number of activations to skip
+   * @param limit maximum number of activations to list
+   * @param includeDocs return document with each activation
+   * @param since timestamp to retrieve activations after
+   * @param upto timestamp to retrieve activations before
+   * @param transid transaction ID for request
+   * @return Future containing a List of activations in JSON format or List of WhiskActivation
```

Comment:
@dubee In which case it would be `JsObject` and not `WhiskActivation`?

Reply:
@chetanmeh I think that's the general list interface currently, where the `Right` is with includeDocs and `Left` not. I don't like that `Either` (no pun intended). But it's due to the design of the ArtifactStore of today. An improvement is very welcome but imho not necessary in this PR.